firewood/src/merkle/node.rs

```diff
@@ -490,6 +495,93 @@ pub enum EncodedNodeType {
     },
 }
 
+// TODO: probably can merge with `EncodedNodeType`.
+#[derive(Debug, Deserialize)]
+struct EncodedBranchNode {
+    chd: Vec<(u64, Vec<u8>)>,
+    data: Option<Vec<u8>>,
+    path: Vec<u8>,
+}
+
+// Note that the serializer passed in should always be the same type as T in EncodedNode<T>.
+impl Serialize for EncodedNode<PlainCodec> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let n = match &self.node {
+            EncodedNodeType::Leaf(n) => {
+                let data = Some(n.data.to_vec()).filter(|d| !d.is_empty());
+                let chd: Vec<(u64, Vec<u8>)> = Default::default();
+                let path = from_nibbles(&n.path.encode(true)).collect();
+                EncodedBranchNode { chd, data, path }
+            }
+            EncodedNodeType::Branch { children, value } => {
+                let mut chd: Vec<(u64, Vec<u8>)> = Default::default();
+
+                for (i, c) in children
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, c)| c.as_ref().map(|c| (i, c)))
+                {
+                    if c.len() >= TRIE_HASH_LEN {
+                        let hash = Keccak256::digest(c).to_vec();
+                        chd.push((i as u64, hash));
+                    } else {
+                        chd.push((i as u64, c.to_vec()));
+                    }
+                }
+
+                let data = value.clone().map(|v| v.0);
+                EncodedBranchNode {
+                    chd,
+                    data,
+                    path: Vec::new(),
+                }
+            }
+        };
+
+        let mut s = serializer.serialize_struct("", 3)?;
+        s.serialize_field("", &n.chd)?;
+        s.serialize_field("", &n.data)?;
+        s.serialize_field("", &n.path)?;
```

Comment:
Can't you just serialize a tuple here?

Reply:
Initially did as a tuple, but I found it is easier to do as a struct without implementation the visitor and directly use `derive`.

Reply:
Plus we probably can just use one struct `EncodedBranchNode` if we only have branch node in the end.